x-pack/plugins/security_solution/server/endpoint/services/actions/action_list.ts

```diff
@@ -0,0 +1,179 @@
+  actionTypes?: string[];
+  context: SecuritySolutionRequestHandlerContext;
+  elasticAgentIds?: string[];
+  endDate?: string;
+  logger: Logger;
+  page?: number;
+  pageSize?: number;
+  startDate?: string;
+  userIds?: string[];
+}
+
+export const getActionList = async ({
+  actionTypes,
+  context,
+  elasticAgentIds,
+  endDate,
+  logger,
+  page: _page,
+  pageSize,
+  startDate,
+  userIds,
+}: GetActionListParam): Promise<ActionListApiResponse> => {
+  const size = Math.floor(pageSize ?? 50 / 2);
+  const page = _page ?? 1;
+  const from = page <= 1 ? 0 : page * size - size + 1;
+
+  const data = await getActionDetailsList({
+    actionTypes,
+    context,
+    elasticAgentIds,
+    endDate,
+    from,
+    logger,
+    size,
+    startDate,
+    userIds,
+  });
+
+  return {
+    page,
+    pageSize,
+    startDate,
+    endDate,
+    userIds,
+    commands: actionTypes,
+    data,
+  };
+};
+
+export interface GetActionDetailsListParam extends GetActionListParam {
+  from?: number;
+  size?: number;
+}
+const getActionDetailsList = async ({
+  actionTypes,
+  context,
+  elasticAgentIds,
+  endDate,
+  from,
+  logger,
+  size,
+  startDate,
+  userIds,
+}: GetActionDetailsListParam): Promise<ActionDetails[]> => {
+  let actions;
+  let actionReqIds;
+
+  try {
+    // fetch actions with matching agent_id
+    const { actionIds, actionRequests } = await getActionsForAgents({
+      actionTypes,
+      context,
+      logger,
+      elasticAgentIds,
+      startDate,
+      endDate,
+      from,
+      size,
+      userIds,
+    });
+    actions = actionRequests;
+    actionReqIds = actionIds;
+  } catch (error) {
+    logger.error(error);
+    throw error;
+  }
+
+  if (actions?.statusCode !== 200) {
+    logger.error(`Error fetching actions log for agent_ids ${elasticAgentIds}`);
+    throw new Error(`Error fetching actions log for agent_ids ${elasticAgentIds}`);
+  }
+
+  // categorize actions as fleet and endpoint actions
+  const categorizedActions = categorizeActionResults({
+    results: actions?.body?.hits?.hits,
+  });
+
+  // normalized actions with a flat structure to access relevant values
+  const normalizedActionRequests: Array<ReturnType<typeof mapToNormalizedActionRequest>> =
+    categorizedActions.map((action) => mapToNormalizedActionRequest(action.item.data));
+
+  // get all responses for given action Ids and agent Ids
+  const actionResponses = await getActionResponses({
+    context,
+    logger,
+    elasticAgentIds,
+    actionIds: actionReqIds,
+  });
+
+  // categorize responses as fleet and endpoint responses
+  const categorizedResponses = categorizeResponseResults({
+    results: actionResponses?.body?.hits?.hits,
+  });
+
+  // compute action details list for each action id
+  const actionDetails: ActionDetails[] = normalizedActionRequests.map((action) => {
+    // pick only those actions that match the current action id
+    const matchedActions = categorizedActions.filter((ac) =>
```

Comment:
Suggestion (here and below) : use a more descriptive variable name other than `ac`